File: services/calculator.js
const COURSE_LIMITS = {
    python: 56, minecraft: 48, roblox: 40, blender: 16,
    scratch: 32, webdesign: 36, webdev: 36,
};

const PAY_RATES = {
    halfLesson: 250,
    smallGroup: 500,
    smallPartial: 300,
    perStudent: 200,
};

function getCourseLimit(groupName) {
    const n = groupName.toLowerCase();
    if (/python/.test(n)) return COURSE_LIMITS.python;
    if (/minecraft/.test(n)) return COURSE_LIMITS.minecraft;
    if (/roblox/.test(n)) return COURSE_LIMITS.roblox;
    if (/blend|блендер/.test(n)) return COURSE_LIMITS.blender;
    if (/scratch/.test(n)) return COURSE_LIMITS.scratch;
    if (/веб.?диз|web.?диз/i.test(n)) return COURSE_LIMITS.webdesign;
    if (/веб.?разр|web.?разр/i.test(n)) return COURSE_LIMITS.webdev;
    return null;
}

function calculatePayment(total, present, isHalf = false) {
    if (present === 0) return 0;
    if (isHalf) return PAY_RATES.halfLesson * present;
    if (total <= 2) {
        return present === total ? PAY_RATES.smallGroup : PAY_RATES.smallPartial;
    }
    return PAY_RATES.perStudent * present;
}

function calculatePenalty(lessonDate, submitDate) {
    // lessonDate и submitDate в формате YYYY-MM-DD
    if (lessonDate === submitDate) return 0;
    return 40; // CONFIG.penaltyAmount
}

function formatMskDate(date = new Date()) {
    return date.toLocaleDateString('ru-RU', {
        timeZone: 'Europe/Moscow',
        year: 'numeric',
        month: '2-digit',
        day: '2-digit'
    }).split('.').reverse().join('-');
}

function formatMskDateTime(date = new Date()) {
    return date.toLocaleString('ru-RU', {
        timeZone: 'Europe/Moscow',
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit'
    });
}

function getWeekStartMsk(now = new Date()) {
    const mskStr = now.toLocaleDateString('ru-RU', { timeZone: 'Europe/Moscow' });
    const [d, m, y] = mskStr.split('.');
    const dt = new Date(y, m - 1, d);
    const dow = dt.getDay();
    const diff = (dow === 0) ? -6 : 1 - dow;
    dt.setDate(dt.getDate() + diff);
    return dt;
}

function mskMonthRange(now = new Date()) {
    const today = formatMskDate(now); // 'YYYY-MM-DD' в МСК
    const [y, m] = today.split('-').map(Number);
    const month = `${y}-${String(m).padStart(2, '0')}`;
    const month_start = `${month}-01`;
    const ny = m === 12 ? y + 1 : y;
    const nm = m === 12 ? 1 : m + 1;
    const month_end = `${ny}-${String(nm).padStart(2, '0')}-01`;
    return { month, month_start, month_end };
}

module.exports = {
    getCourseLimit,
    calculatePayment,
    calculatePenalty,
    formatMskDate,
    formatMskDateTime,
    getWeekStartMsk,
    mskMonthRange,
};

File: journal_django/apps/payroll/calculator.py
"""
calculator.py — расчёт зарплаты преподавателя за урок (payment) и штрафа за
просрочку отчёта (penalty). Общее для teacher SPA и admin SPA — оба пути
записи урока (apps.teacher_spa.services.submit_lesson,
apps.lessons.services.record_lesson) вызывают эти функции, сервер всегда
считает сам (клиентские payment/penalty не принимаются).

Все функции работают с int (рублями), никаких Decimal — оплата у преподавателя
всегда целая (нет копеек).
"""
from __future__ import annotations

PAY_RATES = {
    'halfLesson': 250,    # за каждого присутствующего в полуурочном занятии
    'smallGroup': 500,    # малая группа (1-2 чел.) — все пришли
    'smallPartial': 300,  # малая группа — пришли не все
    'perStudent': 200,    # большая группа (3+ чел.) — за каждого пришедшего
}


def calculate_payment(total: int, present: int, is_half: bool = False) -> int:
    """
    Правила:
      present == 0            → 0
      isHalf                  → 250 * present
      total <= 2, все пришли  → 500
      total <= 2, часть       → 300
      total > 2               → 200 * present
    """
    if present == 0:
        return 0
    if is_half:
        return PAY_RATES['halfLesson'] * present
    if total <= 2:
        return PAY_RATES['smallGroup'] if present == total else PAY_RATES['smallPartial']
    return PAY_RATES['perStudent'] * present


def calculate_penalty(lesson_date: str, submit_date: str, count_students: int) -> int:
    """
    Штраф за просрочку отчёта: тот же день → 0, иначе → 40₽ на каждого
    присутствовавшего ученика. Оба аргумента в формате 'YYYY-MM-DD'.

    Вызывающая сторона решает, что передать в submit_date: teacher SPA — реальную
    сегодняшнюю дату (штраф за опоздание с отчётом); admin SPA передаёт
    submit_date=lesson_date всегда (админ не должен штрафоваться за
    административную запись задним числом — см. design doc).
    """
    if lesson_date == submit_date:
        return 0
    return 40 * count_students
